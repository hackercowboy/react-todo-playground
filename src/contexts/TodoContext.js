import React, {
  createContext, useState, useContext, useMemo,
} from 'react';
import PropTypes from 'prop-types';
import { v4 as uuid } from 'uuid';

const TodoContext = createContext();

function TodoContextProvider({ children }) {
  const [todos, setTodos] = useState([
    {
      id: uuid(), created: new Date(), text: 'Learn React', status: 'pending',
    },
    {
      id: uuid(), created: new Date(), text: 'Bring out the trash', status: 'pending',
    },
    {
      id: uuid(), created: new Date(), text: 'Buy grocery', status: 'pending',
    },
    {
      id: uuid(), created: new Date(), text: 'Mow the lawn', status: 'done',
    },
  ]);

  const context = useMemo(() => {
    const addTodo = (todo) => {
      setTodos([...todos, {
        ...todo, id: uuid(), created: new Date(), status: 'pending',
      }]);
    };

    const deleteTodo = (todo) => {
      setTodos(todos.filter((t) => t.id !== todo.id));
    };

    const finishTodo = (todo) => {
      setTodos(todos.map((t) => (t.id === todo.id ? { ...todo, status: 'done' } : t)));
    };
    return {
      todos, addTodo, deleteTodo, finishTodo,
    };
  }, [todos]);

  return (
    <TodoContext.Provider value={context}>
      {children}
    </TodoContext.Provider>
  );
}

const useTodoContext = () => useContext(TodoContext);

TodoContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export { TodoContext, TodoContextProvider, useTodoContext };
