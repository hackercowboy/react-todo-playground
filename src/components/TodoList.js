import React from 'react';
import TodoListItem from './TodoListItem';
import { useTodoContext } from '../contexts/TodoContext';

import './TodoList.css';

function TodoList() {
  const { todos } = useTodoContext();
  return (
    <ul className="todo-list">
      {todos
        .sort((a, b) => b.created.getTime() - a.created.getTime())
        .sort((a, b) => b.status.localeCompare(a.status)).map((todo) => (
          <TodoListItem key={todo.id} todo={todo} />))}
    </ul>
  );
}

export default TodoList;
