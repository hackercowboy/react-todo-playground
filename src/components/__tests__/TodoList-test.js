import React from 'react';
import { render, prettyDOM } from '@testing-library/react';

import TodoList from '../TodoList';
import { useTodoContext } from '../../contexts/TodoContext';

jest.mock('../../contexts/TodoContext', () => ({
  useTodoContext: jest.fn(),
}));

describe('<TodoList/>', () => {
  it('should render', () => {
    useTodoContext.mockImplementation(() => ({ todos: [{ id: 'abc', status: 'pending' }] }));
    render(<TodoList />);
    expect(prettyDOM()).toMatchSnapshot();
  });
});
