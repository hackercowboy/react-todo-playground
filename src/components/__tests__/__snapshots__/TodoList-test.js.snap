// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList/> should render 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<ul[39m
      [33mclass[39m=[32m\\"todo-list\\"[39m
    [36m>[39m
      [36m<li[39m
        [33mclass[39m=[32m\\"todo-list-item\\"[39m
      [36m>[39m
        [36m<span[39m
          [33mclass[39m=[32m\\"todo-list-item-text pending\\"[39m
        [36m/>[39m
        [36m<span[39m
          [33mclass[39m=[32m\\"todo-list-item-action pending\\"[39m
        [36m>[39m
          [36m<button[39m
            [33mclass[39m=[32m\\"todo-list-item-update-button\\"[39m
            [33mtype[39m=[32m\\"button\\"[39m
          [36m>[39m
            [0mDone[0m
          [36m</button>[39m
        [36m</span>[39m
      [36m</li>[39m
    [36m</ul>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
